### DISCUSSION TITLE
Get the destination routes for an specific path programmatically

### DISCUSSION THREAD

**Discussion by thomasmiko** [dotnet/yarp#550] [Q&A] [answered]
Is it possible to get the destination routes for an specific path programmatically? 

For example, in my configuration I have defined 1 route with 3 destinations:

Route: "/test/{**catchall}"
Destinations: http://server1:1860/destination/*, http://server2:1860/destination/*, http://server3:1860/destination/*

Proxying to the routes is working fine.
But i have one requirement there I must check (within the reverse proxy) if a specific path has at least one destination (and get a list of this destinations):

http://localhost:1860/test/abcd?xyz=123
=> Check should OK, we have 3 destinations

http://localhost:1860/test2/abcd?xyz=123
=> Check should NOT OK, we have no configured destination for /test2/* path

**Comment by Tratcher** [comment, marked as answer]
When do you want to do this check?
- At startup when loading the config
- Per request, checking if this request currently has any destinations
- Some other time, checking if an arbitrary path matches a route and destinations

What do you plan on doing with that list of destinations?

**Reply by thomasmiko** [reply]
Some other time...

We have the requirement that in some situations a caller of the ReverseProxy service should have the option to communicate directly with the target service (performance relevant calls that have to be made to a target service multiple times within a certain period of time).

For example the caller first determines the destination for his route (through the reverse proxy) and can then call the route directly for a specific time.

This is only an exception, but we would need it for some constellations.

**Reply by Tratcher** [reply]
That's a new one...

> performance relevant calls

Have you benchmarked the performance with the proxy? The proxy does add some overhead, but it might still be acceptable. Bypassing a proxy is rarely recommended.

How about this: The caller sends a request to the expected path with a special method, query parameter, header, etc. E.g.
GET /test/ HTTP/1.1
Proxy-Command: ListDestinations
Accept: applicaiton/json

The normal route lookup will run to find the cluster and destinations. You can then have a custom proxy middleware that will intercept requests with that special command and respond with the destination information. Something like this:

```C#
                    proxyPipeline.Use((context, next) =>
                    {
                        if (string.Equals("ListDestinations", context.Request.Headers["Proxy-Command"], StringComparison.OrdinalIgnoreCase))
                        {
                            var proxyFeature = context.Features.Get<IReverseProxyFeature>();
                            var destinations = proxyFeature.AvailableDestinations;
                            return WriteResponse(context, destinations);
                        }

                        return next();
                    });
```

(Compare to this [sample](https://github.com/microsoft/reverse-proxy/blob/25967f49ecb4c55888d4f97ec0f2e19e45bf0f35/samples/ReverseProxy.Code.Sample/Startup.cs#L76).)

**Reply by thomasmiko** [reply]
I've now done some performance tests (with and without proxy) and the overhead is realy low. 
I will discuss it with our product managers, but i think we don't need the bypass anymore. But if we need it, then I will implement the given solution (have tested it in a prototype and it works).
